enterprise/cypher/compiled-expressions/src/test/scala/org/neo4j/cypher/internal/runtime/compiled/expressions/CodeGenerationTest.scala

```diff
@@ -0,0 +1,383 @@
+
+    // Then
+    val value = compiled.compute(ctx, tx, EMPTY_MAP).asInstanceOf[DoubleValue].doubleValue()
+    value should (be > 0.0 and be <= 1.0)
+  }
+
+  test("sin function") {
+    val arg = random.nextDouble()
+    compile(function("sin", literalFloat(arg))).compute(ctx, tx, EMPTY_MAP) should equal(doubleValue(Math.sin(arg)))
+    compile(function("sin", noValue)).compute(ctx, tx, EMPTY_MAP) should equal(NO_VALUE)
+  }
+
+  test("asin function") {
+    val arg = random.nextDouble()
+    compile(function("asin", literalFloat(arg))).compute(ctx, tx, EMPTY_MAP) should equal(doubleValue(Math.asin(arg)))
+    compile(function("asin", noValue)).compute(ctx, tx, EMPTY_MAP) should equal(NO_VALUE)
+  }
+
+  test("haversin function") {
+    val arg = random.nextDouble()
+    compile(function("haversin", literalFloat(arg))).compute(ctx, tx, EMPTY_MAP) should equal((doubleValue((1.0 - Math.cos(arg))/2)))
+    compile(function("haversin", noValue)).compute(ctx, tx, EMPTY_MAP) should equal(NO_VALUE)
+  }
+
+  test("acos function") {
+    val arg = random.nextDouble()
+    compile(function("acos", literalFloat(arg))).compute(ctx, tx, EMPTY_MAP) should equal(doubleValue(Math.acos(arg)))
+    compile(function("acos", noValue)).compute(ctx, tx, EMPTY_MAP) should equal(NO_VALUE)
+  }
+
+  test("cos function") {
+    val arg = random.nextDouble()
+    compile(function("cos", literalFloat(arg))).compute(ctx, tx, EMPTY_MAP) should equal(doubleValue(Math.cos(arg)))
+    compile(function("cos", noValue)).compute(ctx, tx, EMPTY_MAP) should equal(NO_VALUE)
+  }
+
+  test("cot function") {
+    val arg = random.nextDouble()
+    compile(function("cot", literalFloat(arg))).compute(ctx, tx, EMPTY_MAP) should equal(doubleValue(1 / Math.tan(arg)))
+    compile(function("cot", noValue)).compute(ctx, tx, EMPTY_MAP) should equal(NO_VALUE)
+  }
+
+  test("atan function") {
+    val arg = random.nextDouble()
+    compile(function("atan", literalFloat(arg))).compute(ctx, tx, EMPTY_MAP) should equal(doubleValue(Math.atan(arg)))
+    compile(function("atan", noValue)).compute(ctx, tx, EMPTY_MAP) should equal(NO_VALUE)
+  }
+
+  test("tan function") {
+    val arg = random.nextDouble()
+    compile(function("tan", literalFloat(arg))).compute(ctx, tx, EMPTY_MAP) should equal(doubleValue(Math.tan(arg)))
+    compile(function("tan", noValue)).compute(ctx, tx, EMPTY_MAP) should equal(NO_VALUE)
+  }
+
+  test("ceil function") {
+    val arg = random.nextDouble()
+    compile(function("ceil", literalFloat(arg))).compute(ctx, tx, EMPTY_MAP) should equal(doubleValue(Math.ceil(arg)))
+    compile(function("ceil", noValue)).compute(ctx, tx, EMPTY_MAP) should equal(NO_VALUE)
+  }
+
+  test("floor function") {
+    val arg = random.nextDouble()
+    compile(function("floor", literalFloat(arg))).compute(ctx, tx, EMPTY_MAP) should equal(doubleValue(Math.floor(arg)))
+    compile(function("floor", noValue)).compute(ctx, tx, EMPTY_MAP) should equal(NO_VALUE)
```

Comment:
It's great that you are explicitly checking all the null-handling. Lot's of bugs hide in the nulls.